### PR TITLE
feat(filtering): add SimilarityFilterBlock for near-duplicate removal

### PR DESCRIPTION
**Pull request by shivchander**
## Summary

Adds a `SimilarityFilterBlock` to `core/blocks/filtering/` that removes near-duplicate rows from a DataFrame based on text similarity using `difflib.SequenceMatcher` (zero new dependencies).

Original design by **UgaTheDev** (PR #665). This implementation incorporates all review feedback from that closed PR.

### Parameters

| Parameter | Type | Default | Description |
|-----------|------|---------|-------------|
| `input_cols` | `str \| list[str]` | required | Column(s) to compare for similarity (first column is used) |
| `threshold` | `float` | `0.85` | Similarity ratio (0.0–1.0). Rows above this vs any kept row are dropped |
| `group_by` | `str \| None` | `None` | If set, deduplication is scoped within groups of this column |

### Review fixes incorporated from PR #665

1. NaN handling in `group_by`: uses `dropna=False` to preserve rows with missing group keys
2. Empty string similarity: `_similarity("", "")` returns 1.0
3. `input_cols` validator: rejects empty lists via `@field_validator`
4. All tests use `__call__()` path (not `generate()`)
5. Empty DataFrame test asserts `EmptyDatasetError`
6. Missing column test asserts `MissingColumnError`
7. Module docstring describes algorithm as greedy sequential scan
8. No unnecessary `hasattr` guard on `model_post_init`
9. `pydantic` import in correct section
10. Logs when zero duplicates are found
11. Tests for NaN in group_by, NaN in comparison column, first-occurrence retention

### Files changed

| File | Change |
|------|--------|
| `src/sdg_hub/core/blocks/filtering/similarity_filter.py` | New block |
| `src/sdg_hub/core/blocks/filtering/__init__.py` | Added export |
| `src/sdg_hub/core/blocks/__init__.py` | Added export |
| `tests/blocks/filtering/test_similarity_filter.py` | 19 unit tests |

## Tracking

- **GitHub Issue:** Fixes #652
- **Multica Issue:** SDG-22
- **Original PR:** #665 (closed, author unresponsive)

## Test plan

- [x] `uv run pytest tests/structural/` — 138 passed
- [x] `uv run ruff check src/ tests/` — all checks passed
- [x] `uv run mypy src/sdg_hub` — no issues
- [x] `uv run pytest tests/blocks/filtering/ -v` — 37 passed (18 existing + 19 new)
- [x] `uv run pytest tests/blocks tests/connectors tests/flow tests/utils -m "not (examples or slow)"` — 990 passed

## New Features
* Added a Similarity Filter block that removes near-duplicate rows based on text similarity comparison with a configurable threshold (0.0–1.0) and optional per-group deduplication.

[Review Change Stack](https://app.coderabbit.ai/change-stack/Red-Hat-AI-Innovation-Team/sdg_hub/pull/779)

### CODE REVIEW

File: src/sdg_hub/core/blocks/filtering/similarity_filter.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        for idx, row in group.iterrows():
+            text = str(row[col])
+            is_duplicate = any(
+                self._similarity(text, kept) > self.threshold for kept in kept_texts
+            )
+            if not is_duplicate:
+                kept_indices.append(idx)
+                kept_texts.append(text)
+
+        return group.loc[kept_indices]
```
Comment (shivchander) [inline src/sdg_hub/core/blocks/filtering/similarity_filter.py:103]: `group.loc[kept_indices]` uses **label-based** lookup. If the input DataFrame has non-unique index labels (common after `pd.concat` without `reset_index`), `.loc` returns *all* rows matching each label instead of just the intended one, silently re-introducing duplicates.

Fix: switch to positional indexing with `enumerate` + `iloc`:

```python
def _deduplicate_group(self, group: pd.DataFrame, col: str) -> pd.DataFrame:
    kept_positions: list[int] = []
    kept_texts: list[str] = []

    for pos, text in enumerate(group[col].astype(str)):
        is_duplicate = any(
            self._similarity(text, kept) > self.threshold for kept in kept_texts
        )
        if not is_duplicate:
            kept_positions.append(pos)
            kept_texts.append(text)

    return group.iloc[kept_positions]
```

This also avoids the overhead of `iterrows()`. Add a regression test with duplicate index labels to prevent re-introduction.

**Review by shivchander — COMMENTED** [review]
## Code Review — Changes Requested

### Summary
Adds `SimilarityFilterBlock` for near-duplicate row removal using `difflib.SequenceMatcher`. Well-structured, follows the `ColumnValueFilterBlock` pattern, and addresses all 11 review issues from PR #665. 19 tests with good coverage of success, error, and edge cases.

### Issues Found

1. **[similarity_filter.py:105] Correctness — `.loc` vs `.iloc` on non-unique indices**
   `_deduplicate_group` collects index *labels* via `iterrows()` and selects with `.loc`. If the input DataFrame has duplicate index labels (common after `pd.concat` without `reset_index`), `.loc` returns all rows matching each label, silently re-introducing the duplicates the method just removed. Fix: use `enumerate()` + `iloc` for positional indexing. Add a regression test with duplicate-index input.

2. **[CI] `ruff format` failure**
   `similarity_filter.py` needs reformatting. CI already generated a fix branch at `ci-fix/feat/similarity-filter-block`.

### What's Good
- All 11 review items from PR #665 are correctly addressed: NaN-safe groupby, empty-string similarity, input_cols validator, `__call__` path in tests, proper error types, accurate docstring, no hasattr guard, zero-duplicate logging
- Clean separation: `_similarity` is a static pure function, `_deduplicate_group` handles per-group logic, `generate` orchestrates
- Tests are specific and behavioral — assertions check exact row counts, column values, and first-occurrence retention, not just "something was returned"
- Proper Pydantic field validation with `ge`/`le` constraints on threshold
- All unit tests pass across Python 3.10/3.11 on ubuntu and macOS
- 100% codecov patch coverage

### Action Required
1. Switch `_deduplicate_group` from label-based `.loc` to positional `.iloc` (see inline comment for suggested fix)
2. Add a test with a non-unique-index DataFrame as regression guard
3. Apply `ruff format` to fix CI (merge the existing fix branch or run locally)

**Comment by shivchander** [comment]
Closing — redoing through the fixed SDLC (builder → evaluator → lead review → /merge). The underlying GitHub issues remain open and will be re-triaged.
